integration-cli/docker_cli_cp_test.go

```diff
@@ -596,3 +596,82 @@ func (s *DockerSuite) TestCpNameHasColon(c *check.C) {
 		c.Fatalf("Wrong content in copied file %q, should be %q", content, "lololol\n")
 	}
 }
+
+func (s *DockerSuite) TestCpGlob(c *check.C) {
+	command := "mkdir '/other' && mkdir -p '/some/path/other' && echo -n '" + cpContainerContents + "' > " + cpFullPath + " && echo -n '" + cpContainerContents + "' > /some/path/other/test"
+
+	out, exitCode := dockerCmd(c, "run", "-d", "busybox", "/bin/sh", "-c", command)
+	if exitCode != 0 {
+		c.Fatal("failed to create a container", out)
+	}
+
+	cleanedContainerID := strings.TrimSpace(out)
+
+	out, _ = dockerCmd(c, "wait", cleanedContainerID)
+	if strings.TrimSpace(out) != "0" {
+		c.Fatal("failed to set up container", out)
+	}
+
+	if err := os.MkdirAll(cpTestPath, os.ModeDir); err != nil {
+		c.Fatal(err)
+	}
+
+	defer os.RemoveAll(cpTestPathParent)
+
+	tmpdir, err := ioutil.TempDir("", "docker-integration")
+
+	if err != nil {
+		c.Fatal(err)
+	}
+	defer os.RemoveAll(tmpdir)
+
+	cases := []struct {
+		pathMatch string
+		tmpName   []string
+		fail      bool
+	}{
+		{path.Join(cpTestPath, "notfound"), []string{}, true},
+		{path.Join(cpTestPath, "*"), []string{filepath.Join(tmpdir, cpTestName)}, false},
+		{path.Join(cpTestPath, "t[a-z]st"), []string{filepath.Join(tmpdir, cpTestName)}, false},
+		{path.Join(cpTestPath, "*"), []string{filepath.Join(tmpdir, cpTestName), filepath.Join(tmpdir, "other", cpTestName)}, false},
+		{path.Join(cpTestPathParent, "*"), []string{filepath.Join(tmpdir, "path", cpTestName)}, false},
+		{path.Join(cpTestPathParent, "**/*"), []string{filepath.Join(tmpdir, cpTestName), filepath.Join(tmpdir, "other", cpTestName)}, false},
+		{"/*", []string{filepath.Join(tmpdir, "some"), filepath.Join(tmpdir, "other")}, false},
+		{"/", []string{filepath.Join(tmpdir, "some"), filepath.Join(tmpdir, "other")}, false},
+	}
+
+	for _, cs := range cases {
+		out, status, err := runCommandWithOutput(exec.Command(dockerBinary, "cp", cleanedContainerID+":"+cs.pathMatch, tmpdir))
+		if cs.fail {
+			if err == nil {
+				c.Fatalf("Expected error, found %v, status %d\n", out, status)
+			}
+			continue
+		}
+
+		for _, tmpName := range cs.tmpName {
+			fi, err := os.Lstat(tmpName)
+			if err != nil {
+				c.Fatal(err)
+			}
+
+			if fi.IsDir() {
+				continue
+			}
+
+			test, err := ioutil.ReadFile(tmpName)
+			if err != nil {
+				c.Fatal(err)
+			}
+
+			if string(test) == cpHostContents {
+				c.Errorf("output matched host file -- %s - %s\n", cs.pathMatch, cs.tmpName)
+			}
+
+			if string(test) != cpContainerContents {
+				c.Errorf("output doesn't match the input -- %s - %s\n", cs.pathMatch, cs.tmpName)
+			}
+		}
+		os.RemoveAll(filepath.Join(tmpdir, "*"))
```

Comment:
Are you sure "*" is ok in the path? I didn't notice the docs saying it deal with globs.  You might be better off just erasing all of tmpdir and then recreating it before each iteration.